### PR TITLE
Feat/sds 142 delete versions

### PR DESCRIPTION
**Pull request by papamaemae**
## Description of change

Now that object versioning is enabled, a request to delete an object will also need to consider any previous versions of the object. From discussions on supporting GDPR rules, we decided that deleting an object must also delete all its previous versions.

This PR updates the current delete implementation to also delete all versions of that object.

### Summary of changes

**S3 Service Enhancements:**
The service layer has been extended to support listing all versions of a file object and deleting specific versions using their version_id.

**delete_files Endpoint Logic:**
The router logic has been reworked to iterate through and delete all available versions for each provided file_key.

**Audit Service Update:**
The put_item function has been enhanced to include the version_id in the operation history when a file (and its multiple versions) is deleted.

**Testing:**
All unit tests have been updated and extended to reflect the new functionality.
All Postman tests continue to pass; however, they have not been modified or extended, as we are transitioning to Pytest for new end-to-end tests.

**End-to-End Test Restructuring:**
Pytest e2e tests related to the delete_files endpoint have been moved to a new module: test_e2e_delete_files.py. This restructuring groups related delete logic tests together and provides a dedicated space for tests that interact with real S3 operations, allowing assertions on version deletion.

**E2E Helpers Extension:**
The e2e helper utilities have been extended to support listing all versions of an object, enabling more thorough validation in end-to-end tests.

## Link to Jira Ticket

- [SDS-142](https://dsdmoj.atlassian.net/browse/SDS-{ticketNumber})

## Screenshots or test evidence if applicable

Screenshot 2025-09-26 at 00 24 31
Screenshot 2025-09-26 at 00 24 49

### CODE REVIEW

**Review by alanmaydwell — COMMENTED** [review]
Nicely done. 

I've left a few comments. Mainly observations. None are essential but it might be nice to update the one about the test with its own instance of `s3_client` that has mocking disabled. 

By the way, I'm a bit worried that this change has the potential to put more load on the audit table throughput. However, clearly not an issue now as delete is not active in prod, plus we're going to make audit table improvements which should remove this risk.

Also as a bit of extra scrutiny, I created some new files with more than one version, then deleted them via the API. I had a look at the audit table contents these files and they look to have the right sort of thing. 

_Currently a single delete request when a file has more than one version shows as multiple operations in audit table operation_history. I'm not saying this is wrong but it might have the potential to confuse. Might be good to document expected behaviour somewhere_

**Local - via AWS CLI**
Filename 'newversions.md'
```
% awslocal dynamodb get-item --consistent-read --table-name AUDIT_SDS --key '{ "service_id": {"S": "laa-sds-client-local"}, "file_id": {"S": "newversions.md"}}' --region eu-west-1
CREATED_ON      2025-09-26T10:42:59.902604
FILE_ID newversions.md
LAST_UPDATED_ON 2025-09-26T10:44:33.776692
OPERATION_TIME  2025-09-26T10:42:59.902597
OPERATION_TYPE  CREATE
OPERATION_TIME  2025-09-26T10:43:14.220680
OPERATION_TYPE  CREATE
OPERATION_TIME  2025-09-26T10:43:18.080602
OPERATION_TYPE  CREATE
OPERATION_TIME  2025-09-26T10:44:33.721804
OPERATION_TYPE  DELETE
OPERATION_TIME  2025-09-26T10:44:33.751433
OPERATION_TYPE  DELETE
OPERATION_TIME  2025-09-26T10:44:33.776682
OPERATION_TYPE  DELETE
SERVICE_ID      laa-sds-client-local
```

**Dev Environment - via AWS Console**
File `alanfilemultiversion.md`,  then  found the  [relevant audit table row in AWS CLI](https://eu-west-2.console.aws.amazon.com/dynamodbv2/home?region=eu-west-2#item-explorer?filter1Comparator=EQUAL&filter1Name=file_id&filter1Type=S&filter1Value=alanfilemultiversion.md&maximize=true&operation=SCAN&table=cp-01f82b7e7a6205bb) (Note this link does not preserve the filter result - have to re-run)

Operation_history shows as below 

```
[ { "M" : { "OPERATION_TYPE" : { "S" : "CREATE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:21:19.429214" } } }, { "M" : { "OPERATION_TYPE" : { "S" : "CREATE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:21:23.236384" } } }, { "M" : { "OPERATION_TYPE" : { "S" : "CREATE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:21:26.794074" } } }, { "M" : { "OPERATION_TYPE" : { "S" : "DELETE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:25:29.558982" } } }, { "M" : { "OPERATION_TYPE" : { "S" : "DELETE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:25:29.635500" } } }, { "M" : { "OPERATION_TYPE" : { "S" : "DELETE" }, "OPERATION_TIME" : { "S" : "2025-09-26T10:25:29.678834" } } } ]
```
Looks like the right sort of thing.

**Review by alanmaydwell — APPROVED** [review]
Nicely done!

I've approved changes but have also left some comments, some of which suggest possible updates.

File: src/routers/delete_files.py [hunk 1]
```diff
@@ -40,13 +40,35 @@ async def delete_files(
...
+                    logger.info(f"Attempting to delete version with versionId {version_id}")
+                    s3_service.delete_file_version(client_config, file_key, version_id)
+                    audit_service.put_item(
+                        client_config.azure_display_name, file_key, OperationType.DELETE
+                    )
+                    logger.info(f"Deleted version {version_id} of file {file_key}")
+
+                except Exception as e:
+                    logger.error(f"Failed to delete version {version_id} of {file_key}: {e}")
+                    raise e  # Bubble up to outer exception handler
```
[the remark this answers is not in the document]
Comment (papamaemae) [reply src/routers/delete_files.py:69]: Yes, and I wondered about this. Since we want to be able to delete all versions in one go, it makes sense on one hand to throw the error as soon as anything goes wrong.
But I also see the case for collecting errors on a per-version basis, deleting as many as possible and then returning which ones (if any) could not be deleted so only those pesky remaining ones need to be dealt with. Do you think it's worth re-working this for this PR?

File: tests/end_to_end/test_e2e_delete_files.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+@pytest.mark.e2e
+def test_all_versions_of_file_are_deleted():
+    headers = token_getter.get_headers()
+    # Upload multiple versions of a file to be deleted
+    new_filename = make_unique_name("file_to_be_deleted.txt")
+    _ = post_a_file(url=HOST_URL, headers=headers, file_data=test_md_file.get_data(new_filename))
+    _ = post_a_file(url=HOST_URL, headers=headers, file_data=test_md_file.get_data(new_filename))
+    _ = post_a_file(url=HOST_URL, headers=headers, file_data=test_md_file.get_data(new_filename))
+    # Assert that all versions exist in bucket
+    s3_client = LocalS3(mocking_enabled=False)
```
[the remark this answers is not in the document]
Comment (papamaemae) [reply tests/end_to_end/test_e2e_delete_files.py:153]: Thanks for pointing this out. Have now fixed to use global s3_client so test auto-adapts to mocking or 'real' s3 checks based on environment.

File: src/routers/delete_files.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alanmaydwell) [reply src/routers/delete_files.py:69]: I think it's fine to leave it as it is. User would have to make a second attempt if they want to delete all versions, which should pick up all remaining. We can always change later if this turns out to be a problem.

I'd hope this is an edge case anyway and that S3 file deletion is reliable.

**Review by alanmaydwell — APPROVED** [review]
Nicely done and well done for dealing with the tedium of getting the unit test patching right!
